### PR TITLE
Promote Statefulset controller and its e2e tests to use apps/v1

### PR DESCRIPTION
**Pull request by dhilipkumars**
**What this PR does / why we need it**: 
Promotes the statefulset controller to use to use the latest apps group [apps/v1](https://github.com/kubernetes/kubernetes/pull/53679)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes/kubernetes/issues/55714

**Special notes for your reviewer**:

* Listerexpansion for v1 `k8s.io/client-go/listers/apps/v1`  (was recently done for v1beta2)

* `v1beta2` && `v1` had `ObservedGeneration` as `int64` where as `v1beta1` and rest of the code (including conversion) is expecting `ObservedGeneration` to be  `*int64`

```
type StatefulSetStatus struct {
	// observedGeneration is the most recent generation observed for this StatefulSet. It corresponds to the
	// StatefulSet's generation, which is updated on mutation by the API Server.
	// +optional
	ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"varint,1,opt,name=observedGeneration"`
```

* for kubectl's `rollback` and `history` commands a couple functions have been duplicated to allow us to use `v1` version instead of `v1beta1` for statefulsets, while the older functions are still used by other controllers.  

We should be able to remove these duplicates once all the controllers are moved. 

If this aligns with the plan then i could move other controllers too. 

cc: @kow3ns 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dhilipkumars** [comment]
/sig apps
@kubernetes/sig-apps-api-reviews

**Comment by ncdc** [comment]
Do we need to worry about this potentially breaking GKE (if GKE doesn't have `apps/v1` yet)?

**Comment by dhilipkumars** [comment]
@ncdc 
_"Do we need to worry about this potentially breaking GKE (if GKE doesn't have apps/v1 yet)?"_

This seems to paas e2e-gke-gci now is it good enough?

**Review by kow3ns — COMMENTED** [review]
1. The title of this PR is confusing. StatefulSet is promoted to v1. What you are doing is converting the controller to use the v1 client. 
2. You can't do this now. If we merge this PR we break cluster upgrade for any cluster with multiple masters, and rolling back an upgrade becomes hazardous. We can accept a PR like this for the 1.10 release.

**Comment by dhilipkumars** [comment]
@kow3ns Sure, I can maintain this PR until 1.9 is cut out of `master`, in the meanwhile could we mark this to `1.10` milestone, please?

**Comment by kow3ns** [comment]
@dhilipkumars don't close it. We will do this in the 1.10 timeframe and we can review and merge when we come out of code freeze.

**Comment by mattfarina** [comment]
@dhilipkumars Now that we're working on 1.10, should this get ready for that?

**Comment by dhilipkumars** [comment]
@mattfarina  Agreed.  I will fix the CI in the morning, should be ready for review.

**Comment by dhilipkumars** [comment]
@mattfarina & @kow3ns PTAL, is all green now.

**Comment by antoineco** [comment]
Should this also include the management of Conditions (added in #55268)?

**Comment by dhilipkumars** [comment]
@antoineco Conditions are already a part of v1 api [object](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/apps/v1/types.go#L220)? Is there a specific functionality we should include when you say 'condition management' ?

**Comment by antoineco** [comment]
@dhilipkumars I meant actually computing the resource conditions at the end of the reconciliation loop. Currently it seems like the `sts` AP can take conditions, but nothing is generating them.

**Comment by dhilipkumars** [comment]
@antoineco I'm not sure that is the scope of this PR, this is simply making the core controller work on top of `v1` type objects instead of v1beta.. objects.

**Comment by dhilipkumars** [comment]
ping @kubernetes/sig-apps-api-reviews

File: pkg/apis/apps/v1/conversion.go [hunk 1]
```diff
@@ -453,8 +453,10 @@ func Convert_apps_StatefulSetUpdateStrategy_To_v1_StatefulSetUpdateStrategy(in *
 }
 
 func Convert_v1_StatefulSetStatus_To_apps_StatefulSetStatus(in *appsv1.StatefulSetStatus, out *apps.StatefulSetStatus, s conversion.Scope) error {
-	out.ObservedGeneration = new(int64)
-	*out.ObservedGeneration = in.ObservedGeneration
+	if in.ObservedGeneration != nil {
```
Comment (dixudx) [inline pkg/apis/apps/v1/conversion.go]: This does not make sense.
`in. ObservedGeneration` is `int64`, which cannot be `nil`.
I prefer to keep this unchanged.

File: pkg/apis/apps/v1/conversion.go [hunk 2]
```diff
@@ -476,7 +478,8 @@ func Convert_v1_StatefulSetStatus_To_apps_StatefulSetStatus(in *appsv1.StatefulS
 
 func Convert_apps_StatefulSetStatus_To_v1_StatefulSetStatus(in *apps.StatefulSetStatus, out *appsv1.StatefulSetStatus, s conversion.Scope) error {
 	if in.ObservedGeneration != nil {
-		out.ObservedGeneration = *in.ObservedGeneration
+		out.ObservedGeneration = new(int64)
```
Comment (dixudx) [inline pkg/apis/apps/v1/conversion.go]: `out.ObservedGeneration` is type `int64`. Should not create a new int64 ptr here.
Keep this unchanged.

Actually there are inconsistence between `apps/internal` and `apps/v1`. Such as `ObservedGeneration` is type `*int64` in `apps/internal`, while it is changed to `int64` in `apps/v1`.

**Review by dixudx — CHANGES_REQUESTED** [review]

File: pkg/apis/apps/v1/conversion.go (hunk 1 above)
Comment (dhilipkumars) [reply pkg/apis/apps/v1/conversion.go]: @dixudx [v1beta1](https://github.com/kubernetes/kubernetes/blob/091794c0874bed681cdf221dd7fbc8f31dbeb0b1/staging/src/k8s.io/api/apps/v1beta1/types.go#L222) still uses *int64.  It is inconsistent across api versions which one should we follow?

Comment (dixudx) [reply pkg/apis/apps/v1/conversion.go]: Refer to #49607. Don't change the type back.

File: pkg/apis/apps/v1/conversion.go [hunk 3]
```diff
@@ -453,6 +453,7 @@ func Convert_apps_StatefulSetUpdateStrategy_To_v1_StatefulSetUpdateStrategy(in *
 }
 
 func Convert_v1_StatefulSetStatus_To_apps_StatefulSetStatus(in *appsv1.StatefulSetStatus, out *apps.StatefulSetStatus, s conversion.Scope) error {
+
```
Comment (dixudx) [inline pkg/apis/apps/v1/conversion.go]: Remove this newline.

File: pkg/apis/apps/v1/conversion.go [hunk 4]
```diff
@@ -472,9 +473,11 @@ func Convert_v1_StatefulSetStatus_To_apps_StatefulSetStatus(in *appsv1.StatefulS
 		}
 	}
 	return nil
+
```
Comment (dixudx) [inline pkg/apis/apps/v1/conversion.go]: Ditto.

File: pkg/apis/apps/v1/conversion.go [hunk 5]
```diff
@@ -472,9 +473,11 @@ func Convert_v1_StatefulSetStatus_To_apps_StatefulSetStatus(in *appsv1.StatefulS
...
 }
 
 func Convert_apps_StatefulSetStatus_To_v1_StatefulSetStatus(in *apps.StatefulSetStatus, out *appsv1.StatefulSetStatus, s conversion.Scope) error {
+
```
Comment (dixudx) [inline pkg/apis/apps/v1/conversion.go]: Ditto.

File: pkg/controller/statefulset/stateful_set_control.go [hunk 6]
```diff
@@ -22,7 +22,8 @@ import (
 
 	"github.com/golang/glog"
 
-	apps "k8s.io/api/apps/v1beta1"
+	apps "k8s.io/api/apps/v1"
+	//apps "k8s.io/api/apps/v1beta1"
```
Comment (dixudx) [inline pkg/controller/statefulset/stateful_set_control.go]: Forgot to remove this?

File: pkg/controller/statefulset/stateful_set_utils.go [hunk 7]
```diff
@@ -23,7 +23,8 @@ import (
 	"regexp"
 	"strconv"
 
-	apps "k8s.io/api/apps/v1beta1"
+	//apps "k8s.io/api/apps/v1beta1"
```
Comment (dixudx) [inline pkg/controller/statefulset/stateful_set_utils.go]: Forgot to remove this ?

File: pkg/kubectl/rollback.go [hunk 8]
```diff
@@ -338,13 +339,39 @@ func (r *StatefulSetRollbacker) Rollback(obj runtime.Object, updatedAnnotations
...
+	if _, err = r.c.AppsV1().StatefulSets(sts.Namespace).Patch(sts.Name, types.StrategicMergePatchType, toHistory.Data.Raw); err != nil {
 		return "", fmt.Errorf("failed restoring revision %d: %v", toRevision, err)
 	}
 
 	return rollbackSuccess, nil
 }
 
+// findHistoryV1 returns a controllerrevision of a specific revision from the given controllerrevisions.
+// It returns nil if no such controllerrevision exists.
+// If toRevision is 0, the last previously used history is returned.
```
Comment (dixudx) [inline pkg/kubectl/rollback.go]: Add a TODO here to rename this to findHistory when other controllers have been upgraded.

File: test/e2e/apps/statefulset.go [hunk 9]
```diff
@@ -23,8 +23,9 @@ import (
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
-	apps "k8s.io/api/apps/v1beta1"
-	appsv1beta2 "k8s.io/api/apps/v1beta2"
+	apps "k8s.io/api/apps/v1"
+	//appsv1beta1 "k8s.io/api/apps/v1beta1"
```
Comment (dixudx) [inline test/e2e/apps/statefulset.go]: Remove these two comment lines.

File: test/e2e/apps/statefulset.go [hunk 10]
```diff
@@ -844,7 +845,7 @@ var _ = SIGDescribe("StatefulSet", func() {
 				return nil
 			}, framework.StatefulPodTimeout, 2*time.Second).Should(BeNil())
 		})
-
+		/* Comment it for now.
```
Comment (dixudx) [inline test/e2e/apps/statefulset.go]: Any reasons to comment it?

**Comment by dhilipkumars** [comment]
cc: @smarterclayton could you please approve.

**Pull request merged**